docs/developer-guide/serialization.rst

```diff
@@ -0,0 +1,269 @@
+.. _ksql_serialization:
+
+KSQL Serialization
+==================
+
+KSQL currently supports three serialization formats:
+
+*. ``DELIMITED`` supports comma separated values. See :ref:`delimited_format` below.
+*. ``JSON`` supports JSON values. See :ref:`json_format` below.
+*. ``AVRO`` supports AVRO serialized values. See :ref:`avro_format` below.
+
+
+.. _delimited_format
+
+---------
+DELIMITED
+---------
+
+The ``DELIMITED`` format supports comma separated values.
+
+The serialized object should be a Kafka serialized string, which will be split into columns.
+
+For example, given a KSQL statement such as:
+
+.. code:: sql
+
+    CREATE STREAM x (ID BIGINT, NAME STRING, AGE INT) WITH (VALUE_FORMAT='DELIMITED', ...);
+
+Then KSQL would split a value of ``120, bob, 49`` into the three fields with ``ID`` of ``120``,
+``NAME`` of ``bob`` and ``AGE`` of ``49``.
+
+This data format supports all KSQL data types except ``ARRAY``, ``MAP`` and ``STRUCT``. See
+:ref:`data-types` for more details.
+
+.. _json_format
+
+----
+JSON
+----
+
+The ``JSON`` format supports JSON values.
+
+The JSON format supports all of KSQL's data types. (See :ref:`data-types` for details). Though,
+as JSON itself does not support a map type, KSQL serializes ``MAP``s as JSON objects.  Because of
+this the JSON format can only support ``MAP``s with ``STRING`` keys.
+
+The serialized object should be a Kafka serialized string containing a valid JSON value. The foramt
+supports JSON objects and top-level primitives, arrays and maps. See below for more info.
+
+JSON Objects
+------------
+
+Values that are JSON objects are probably the most common.
+
+For example, given a KSQL statement such as:
+
+.. code:: sql
+
+    CREATE STREAM x (ID BIGINT, NAME STRING, AGE INT) WITH (VALUE_FORMAT='JSON', ...);
+
+And a JSON value of:
+
+.. code:: json
+
+       {
+         "id": 120,
+         "name": "bob",
+         "age": "49"
+       }
+
+KSQL would deserialize the JSON object's fields into the corresponding field of the stream.
+
+Top-level Primitives
+--------------------
+
+The JSON format supports reading top-level JSON primitives. KSQL can only do so if the target
+schema contains only a single field of a compatible type.
+
+For example, given a KSQL statement with only a single field in the value schema:
+
+.. code:: sql
+
+    CREATE STREAM x (ID BIGINT) WITH (VALUE_FORMAT='JSON', ...);
+
+And a JSON value of:
+
+.. code:: json
+
+       10
```

Comment:
How do we distinguish between 

```JSON
10
```
and 
```JSON
{ "ID":"10"}
```
I assumed we were going to add new formats, `JSON_PRIMITIVE` and `AVRO_PIRIMITIVE`. Or maybe something like `JSON_RAW` and `AVRO_RAW` since we can have ARRAY or MAP too.
This would indicate that the value is not a JSON document or Avro recored.

Reply:
I distinguish between the two by virtue of the fact they are different ;)

`FORMAT='JSON'` is enough on the deserialize side as you have the data. If the value deserializes to a primitive, then we treat it as a primitive. If we deserial to a JSON object or Avro record then we treat it as such.